api4/group.go

```diff
@@ -767,6 +773,28 @@ func getGroups(c *Context, w http.ResponseWriter, r *http.Request) {
 		return
 	}
 
+	// if the team is group-constrained, only display the intersect of the teams
+	// groups and the groups returned by the query.
+	if parentTeam != nil && parentTeam.IsGroupConstrained() {
```

Comment:
A thought: because we do the paging logic prior to the intersection, are we sure the result will be as expected? Eg, if you have 20 groups, Page = 0, PerPage = 10, but only one group is in the intersection, you'll get one group in the response, even though there may be more in the next page -- the user will think "ok, I guess only one group in total" not thinking to call for Page = 1.

Reply:
You're right. This change is breaking pagination. I hadn't noticed, so thank you. This either needs to incorporate the change into the data tier or request the next page(s) of results—by making further DB requests—to fill in any results removed so that the client can know whether there will be more results in the next page or not. I'm going to make one of those updates.

Reply:
@cpoile I'm going to switch to this: https://github.com/mattermost/mattermost-server/pull/14619/files#diff-2c6106afe8477623894c02707bffe06dR1219-R1238. I'll request a re-review once all of the additional tests—including covering pagination—are in place.